providers/google.go

```diff
@@ -179,21 +179,53 @@ func getAdminService(adminEmail string, credentialsReader io.Reader) *admin.Serv
 }
 
 func userInGroup(service *admin.Service, groups []string, email string) bool {
+	// Check the user is part of the group list or one of their subgroups
+	// Note: the following loop won't work for users not having the same domain than the group
+	// There is a domain scope issue on Google side, community is still waiting for an answer from them. See the partial fallback after this loop
+	// https://issuetracker.google.com/issues/109861216
 	for _, allowedgroup := range groups {
 		resp, err := service.Members.HasMember(allowedgroup, email).Do()
 
 		if err != nil {
-			log.Printf("Error calling service.Members.HasMember(%s, %s): %s", allowedgroup, email, err)
-			return false
-		}
-
-		if resp.IsMember {
+			// We do not fail on error since it can come from the domain scope issue mentioned above
+			// Like that we are able to try the fallback implementation
```

Comment:
I think it would be much better if we could determine up-front whether this email address belongs to this google account's domains or not. Maybe one api request up-front can determine this? Then we can fail fast here for members of this account's domains, and jump straight to listing groups for other addresses.

Reply:
I'm not sure we can avoid requesting Google APIs because guess the user wants to connect with this email:
`thomas@domain.io`

And that the proxy has been configured to detect `.com` extension like that:
`-google-group=group-test@domain.com`

We cannot be aware that `domain.io` is an alias of `domain.com`.

Reply:
Since each domain the admin allows can have different "scope domain", we have to loop over each group entries even if we were willing to get all aliases.

Example:
```
-google-group=group-test-1@company-dept-1.com
-google-group=group-test-2@company-dept-2.com
...
```

Tell me if I'm wrong, but after getting aliases (first request) we have to call `hasMember` (second request) to be sure the member is part of the group.

A possible way of saving some requests would be if the proxy looks for some groups with the same domain. So if one of them fails, we probably knows it's gonna be the same for the other (if at least the admin has specified them with the same domain, and not different aliases). But I also see a problem with that, Google responds with a 400 error with a message that could be modified in the future, there is no Error code that specifies `hasMember` error is due to different domain scopes. We would have to consider the english sentence to trigger the flag "hey, we catched the right error".

What do you think?

Reply:
I was not suggesting to avoid making a request to the google API - I'd like to find some single api request we can make in order to know that an email is or is not part of the organization. Maybe there's some api that can give us one of these things:
  * all domains and aliases for this google org
  * whether a domain is a domain or alias for this google org
  * whether this address is a member of this google org

Then we can do only the correct loop of requests, for this member or non-member, and return/log errors better.

Reply:
I think matching on the text in the error response is OK for this purpose.

Also I don't think we should try to cover every theoretically possible case ... if we can cover a few different useful setups with a reasonably small amount of complexity, that would be good. If you want to use oauth2_proxy you can't have any arbitrarily complex setup, it has to be reasonable, but you do have a few options to make it work for you.

This is also why I'm not treating this very urgently, what I would consider to be the primary case is already pretty well supported (all addresses are members of the "domain scope"). If we can extend this just a bit and support addresses that are clearly/obviously not members of the "domain scope" or organization at all, I think that would support the other most common use-case.